File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = []
    let map = {']': '[', ')': '(', '}': '{'}
    
    for(const c of s) {
        if (c in map) {
            if (stack.length == 0 || stack[stack.length - 1] != map[c]) return false
            else stack.pop()
        } else {
            stack.push(c)
        }
    }
    
    return stack.length == 0
};

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stk = new Stack<>();
        Map<Character, Character> pairs = Map.of(
            ')', '(',
            '}', '{',
            ']', '['
        );
        
        for(char c : s.toCharArray()) {
            if (pairs.containsKey(c)) {
                if (stk.isEmpty() || stk.peek() != pairs.get(c)) return false;
                stk.pop();
            } else {
                stk.push(c);
            }
        }
        
        return stk.isEmpty();
    }
}